packages/grid/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx

```diff
@@ -101,13 +101,20 @@ export const useGridColumnHeaders = (props: UseGridColumnHeadersProps) => {
   const rootProps = useGridRootProps();
   const innerRef = React.useRef<HTMLDivElement>(null);
   const handleInnerRef = useForkRef(innerRefProp, innerRef);
-  const [renderContext, setRenderContext] = React.useState<GridRenderContext | null>(null);
+  const [renderContext, setRenderContextRaw] = React.useState<GridRenderContext | null>(null);
   const prevRenderContext = React.useRef<GridRenderContext | null>(renderContext);
   const prevScrollLeft = React.useRef(0);
   const currentPage = useGridVisibleRows(apiRef, rootProps);
   const totalHeaderHeight = getTotalHeaderHeight(apiRef, rootProps.columnHeaderHeight);
   const headerHeight = Math.floor(rootProps.columnHeaderHeight * densityFactor);
 
+  const setRenderContext = React.useCallback((nextRenderContext: GridRenderContext | null) => {
+    if (renderContext && nextRenderContext && areRenderContextsEqual(renderContext, nextRenderContext)) {
+      return;
+    }
+    setRenderContextRaw(nextRenderContext);
+  }, [renderContext])
```

Comment:
We can potentially avoid getting the `renderContext` from closure:

```suggestion
  const setRenderContext = React.useCallback((nextRenderContext: GridRenderContext | null) => {
    setRenderContextRaw(prevRenderContext => {
      if (prevRenderContext && nextRenderContext && areRenderContextsEqual(prevRenderContext, nextRenderContext)) {
        return prevRenderContext;
      }
      return nextRenderContext;
    });
  }, [])
```

Not sure if this matters in this case though